SIPS/sip-7.md

```diff
@@ -2,307 +2,205 @@
-  | string
-  | Json[]
-  | { [prop: string]: Json };
-```
+A Surface is place in the wallet UI in which a snap can expose it's own components.
 
-### Components
+> Example Surfaces would be a dialog window, [transaction insights](./sip-3.md), creating an account flow.
 
-This SIP proposes that Snaps compose their own UIs using the existing MetaMask components, but for the sake of simplicity and security, we only allow this composition via JSON.
+An Action is an external change in the UI system, started either by the user, or the snap.
 
-A user interface definition is therefore a JSON object composed of different types of components, which on render will be converted to React by the MetaMask extension.
+> Example Actions would be change of text input, press of a button, request to close the dialog.
 
-All components will derive from a base component and have at a minimum a `type`.
+## Components
 
-```tsx
-interface BaseComponent {
-  type: string;
-};
-```
+An interface created by a snap SHALL be declarative in form of a JSON-serializable JavaScript value.
 
-### Initial components
+## Base Components
 
-This SIP describes a couple of components for the initial version of the UI:
+All component types inside UI declaration SHALL be derived from either `Parent` or `Literal`.
 
-1. A panel component, which renders a list of `children` components.
-2. A heading component, rendering the text as a HTML heading (`h1`, `h2`, `h3`, and so on).
-    1. The heading is rendered dynamically based on the nesting of panels, i.e.:
-        1. A heading in a panel at the root SHOULD render a `h1`.
-        2. A heading in a panel nested in the root panel SHOULD render a `h2`.
-        3. etc.
-3. A plain text component, rendering the text as a HTML paragraph (`p`).
-4. A button component, rendering the text as a HTML button (`button`), with an optional ID for handling user input.
-    1. The button can have two variants: `primary` and `secondary`.
-    2. If the variant is not specified, `primary` SHOULD be used.
-5. A divider component, rendering a HTML thematic break (`hr`).
-6. A spacer component, pushing content around it to the top and bottom, similar to how `flex-grow` works.
-7. A spinner component, rendering a spinner to indicate loading data.
+### Node
 
-```tsx
-type ComponentType = 'panel' | 'heading' | 'text' | 'button' | 'divider' | 'spacer' | 'spinner';
+The base type for all the components.
 
-interface PanelComponent extends BaseComponent {
-  type: 'panel';
-  children: Component[];
+```typescript
+interface Node {
+  type: string;
 }
+```
 
-interface HeadingComponent extends BaseComponent {
-  type: 'heading';
-  text: string;
-}
+### Parent
 
-interface TextComponent extends BaseComponent {
-  type: 'text';
-  text: string;
-}
-
-interface ButtonComponent extends BaseComponent {
-  type: 'button';
-  variant?: 'primary' | 'secondary';
-  text: string;
-  name?: string;
+```typescript
+interface Parent extends Node {
+  children: Node[];
 }
```

Comment:
I don't think we need a separate `Parent` interface. Nodes that can have children can just define that in their spec.

Reply:
The whole structure was inspired by [Unist](https://github.com/syntax-tree/unist). Having defined Parent and Literal brings common fields so that automatic tools can operate on the whole UI structure without needing to understand each Node separately.

An example of such tool is [unist-util-visit](https://github.com/syntax-tree/unist-util-visit#visittree-test-visitor-reverse), which is extremely useful and [used extensively](https://github.com/MetaMask/SIPs/blob/60c212d1da607005bef989fd364565f21c66f111/tools/validate/src/rules/bad-link.ts#L86) in SIP validation tool and other local parser experiments I've build when playing with UI.